### PR TITLE
Add Fully Convolutional Change Detection (FCCD) models

### LINKED ISSUE #107: Add baseline change detection models
As we are adding more change detection (CD) datasets to torchgeo, we should look to add standard baseline CD models with pretrained weights.

I've implemented some standard CD models ([FCEF, FSiamConc, FCSiamDiff](https://github.com/isaaccorley/torchrs/blob/main/torchrs/models/fc_cd.py) proposed in [1]) in torchrs that we can easily pull into torchgeo. They are essentially modified U-Nets implemented in pure PyTorch with the exception of using `einops.rearrange` in place of some permute, transpose, flatten ops for readability (not sure if we want to add einops as an additional requirement, it seems to be very lightweight with no dependencies).

image

These models are commonly used as baseline comparisons for recent change detection papers [2], [3] and still provide great performance on benchmark datasets.

References:
1. [Fully Convolutional Siamese Networks for Change Detection](https://arxiv.org/abs/1810.08462)
2. [Remote Sensing Image Change Detection with Transformers](https://arxiv.org/abs/2103.00208)
3. [Self-supervised pre-training enhances change detection in Sentinel-2 imagery](https://arxiv.org/abs/2101.08122)

### PR DESCRIPTION
**Pull request by isaaccorley**
- Adds `FCEF`, `FCSiamConc`, `FCSiamDiff` to `torchgeo.models`.
- Adds `einops` as a dependency
- Added unit tests `tests/models/tests_fccd.py`
- Modified all models to work with any `t` (in the case of multitemporal change detection where `t>2`)

image

Example:

```python
import torch
from torchgeo.models import FCEF, FCSiamDiff, FCSiamConc

b, t, c, h, w = 2, 2, 3, 64, 64
x = torch.randn(b, t, c, h, w)

model = FCEF(in_channels=3, t=2, classes=2)
model(x)  # (b, classes, h, w)

model = FCSiamConc(in_channels=3, t=2, classes=2)
model(x)  # (b, classes, h, w)

model = FCSiamDiff(in_channels=3, t=2, classes=2)
model(x)  # (b, classes, h, w)
```

Reference:
- [Paper: "Fully Convolutional Siamese Networks for Change Detection", Daudt et al. (2018)](https://arxiv.org/abs/1810.08462)
- [Official author implementation](https://github.com/rcdaudt/fully_convolutional_change_detection)

Closes #107

### CODE REVIEW

File: spack.yaml [hunk 1]
```diff
@@ -4,6 +4,7 @@ spack:
   - "pil@2.9:"
   - "python@3.6:+bz2"
   - "py-black@21:"
+  - "py-einops"
```
Comment (adamjstewart) [inline spack.yaml:7]: Looks like we don't yet have a Spack package for `py-einops`, I'll add one.

**Review by adamjstewart — COMMENTED** [review]
These look really interesting. @calebrob6 do you know if any of these models are already implemented in any of the model libraries we depend on?

I'm always hesitant to add new dependencies to TorchGeo. As you've probably noticed, we have quite a few dependencies already, and that number seems to increase with every PR. One way in which we've handled this is to make these dependencies optional using lazy imports. This is useful, for example, in a dataset like VHR-10, which requires `rarfile` and `pycocotools`, even though no other dataset requires these. Torchvision does the same thing with many of their datasets. We may want to consider doing something similar here since only a single set of models requires it. @calebrob6 thoughts?

File: spack.yaml (hunk 1 above)
Comment (isaaccorley) [reply spack.yaml:7]: My fault, I assumed there was one.

**Comment by isaaccorley** [comment]
I'm indifferent on whether we use einops or not. However, I've seen a lot of really unreadable remote sensing paper model code using chained `permute`, `reshape`, etc. ops, particularly with attention/transformer based image models, so I'm quite bullish on einops being the right way to implement models. It's a great way to avoid phantom bugs in models.

**Comment by isaaccorley** [comment]
I looked at the dependencies and I don't believe these models are implemented in any of them. I think if we wanted to use pretrained models we may choose to modify a U-Net encoder/decoder from `segmentation_models_pytorch` which is totally fine imo. I only implemented it without another library to stay true to the paper.

File: requirements.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 # black 21+ required for Python 3.9 support
 black>=21.4b0
+einops
```
Comment (adamjstewart) [inline requirements.txt:3]: Can you use pip to install successively older versions of einops and see what the minimum required version is for the tests to pass?

Comment (adamjstewart) [reply requirements.txt:3]: Also, can you add this dependency to `docs/requirements.txt` as well? Unfortunately we have to duplicate that file due to https://github.com/pytorch/pytorch_sphinx_theme/issues/143

Comment (isaaccorley) [reply requirements.txt:3]: In this case, do we want to just leave it w/o a minimum required version?

Comment (adamjstewart) [reply requirements.txt:3]: Yes, just leave it without then

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
After this PR I'm seeing the following pytest warnings:
```
tests/models/test_fccd.py::TestFCEF::test_in_channels[1-1]
  /opt/hostedtoolcache/Python/3.9.6/x64/lib/python3.9/site-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
    return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
```
Any idea what could be causing this?

**Comment by isaaccorley** [comment]
Yes, it's due to the model containing MaxPool2D layers. It's not something we seem to have control over though as it's under the hood in PyTorch. My guess is it will disappear in the next PyTorch update.

**Comment by adamjstewart** [comment]
Ah cool, looks like it has already been fixed in develop and they're just waiting for a new release: https://github.com/pytorch/pytorch/pull/60059

**Pull request merged**